cdc-sandbox/db/shrink.sql

```diff
@@ -0,0 +1,885 @@
+-- THE FOLLOWING SQL WILL SHRINK ANY DATABASE TO AN ARBITRARY NUMBER OF PATIENTS
+
+-- The physical size of the current production anonymized masters are too unwieldy (restore time, size, etc.). 
+-- This script pares down any database to an arbitrary size with the most performant method. 
+-- Since tables are very large any method using deletes would take forever to run. 
+-- So we chose to rebuild from the ground up while satisfying any constraints 
+-- (that the subset of patients requires) in the most performant manner as possible as follows:
+
+-- 1) analyze all schemas and identify largest schemas/tables and start with biggest and work our way down
+-- 2) for each schema remove all database constraints
+-- 3) rebuild tables without indexes (while maintaining copies of the old and creating new unique key names to prevent collisions)
+-- 4) bulk insert with select intos from old tables (fast since no indexes)
+-- 5) re-add/build indexes (with new unique names to avoid collisions)
+-- 6) rename tables (switch old and new)
+-- 7) remove old tables
+
+-- THESE WERE THE SIZES FOR EACH SCHEMA BASED ON A FULL CDC ANONYMIZED DB:
+
+-- NBS_DataIngest 16M
+-- NBS_MSGOUTE 153M
+-- NBS_ODSE 1.6T <- EXTREMELY LARGE
+-- NBS_SRTE 285M
+-- RDB 1.6T <- EXTREMELY LARGE
+
+-- FOLLOWING QUERY LISTS LARGEST TABLE SIZES IN THE SCHEMA:
+  select schema_name(tab.schema_id) + '.' + tab.name as [table],
+    cast(sum(spc.used_pages * 8)/1024.00 as numeric(36, 2)) as used_mb,
+    cast(sum(spc.total_pages * 8)/1024.00 as numeric(36, 2)) as allocated_mb
+from sys.tables tab
+    inner join sys.indexes ind 
+        on tab.object_id = ind.object_id
+    inner join sys.partitions part 
+        on ind.object_id = part.object_id and ind.index_id = part.index_id
+    inner join sys.allocation_units spc
+        on part.partition_id = spc.container_id
+group by schema_name(tab.schema_id) + '.' + tab.name
+order by sum(spc.used_pages) desc
+
+-- LARGEST TABLES IN NBS_ODSE:
+-- |table                                  |used_mb  |allocated_mb|
+-- |---------------------------------------|---------|------------|
+-- |dbo.Observation                        |66,942.38|66,943.86   |
+-- |dbo.Participation                      |54,708.93|54,729.25   |
+-- |dbo.Entity_id                          |49,674.71|49,676.54   |
+-- |dbo.Person                             |46,704.04|46,707.88   |
+-- |dbo.Entity_locator_participation       |39,826.88|39,828.16   |
+-- |dbo.Act_id                             |33,323.13|33,323.6    |
+-- |dbo.Entity_loc_participation_hist      |28,392.43|28,392.55   |
+-- |dbo.Act_relationship                   |27,324.23|27,327.36   |
+-- |dbo.Person_name                        |23,338.17|25,214.56   |
+
+USE NBS_ODSE;
+
+-- Disable constraints for all tables in the database:
+EXEC sp_msforeachtable 'ALTER TABLE ? NOCHECK CONSTRAINT ALL'
+
+CREATE TABLE NBS_ODSE.dbo.Person_shrunken (
```

Comment:
Each Patient will have a row in the `Entity` table with a `class_cd` of `PSN` and an `entity_uid` equal to  `person.person_uid`.

Reply:
Ok is there something I need to do on this one?  Entity table wasn't shrunken because it was about 26th in the list of table sizes and not too large.

Reply:
There is an `Entity` row for each Patient, if we are removing patients we should remove the `Entity` also.  Not sure if leaving the rows in the `Entity` table will have an impact or not.